### PR TITLE
Constrain app bar header paint origin on iOS overscroll

### PR DESCRIPTION
**Pull request by cbracken**
Prevent the app bar header from being overscrolled on iOS by setting
paintOrigin to the overlap.

### CODE REVIEW

**Comment by cbracken** [comment]
/cc @abarth 

Fixes #8466. Thinking up tests for iOS overscroll.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 1]
```diff
@@ -209,6 +209,7 @@ abstract class RenderSliverScrollingPersistentHeader extends RenderSliverPersist
     final double paintExtent = maxExtent - constraints.scrollOffset;
     geometry = new SliverGeometry(
       scrollExtent: maxExtent,
+      paintOrigin: constraints.overlap,
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart]: You might want to do this only when the overlap is negative.  Consider the case when you have a pinned header followed by this kind of header.  You want this kind of header to scroll under the pinned header.  The way you've written this, this header will shift its paint location by the overlap between the two.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 2]
```diff
@@ -271,6 +272,7 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
     final double layoutExtent = maxExtent - constraints.scrollOffset;
     geometry = new SliverGeometry(
       scrollExtent: maxExtent,
+      paintOrigin: constraints.overlap,
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart]: Same comment here.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart (hunk 1 above)
Comment (cbracken) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart (hunk 2 above)
Comment (cbracken) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart]: Done.

**Comment by abarth** [comment]
Looks plausible except its missing tests.

**Comment by cbracken** [comment]
Getting familiar with flutter rendering tests -- should have something later this evening or tomorrow morning I suspect :)

**Comment by abarth** [comment]
You can also test it from the widgets layer.  Take a look at the test I wrote when I added this to the pinned app bar.

**Comment by cbracken** [comment]
Thanks for the pointer! Added.

**Comment by abarth** [comment]
LGTM

**Comment by cbracken** [comment]
Went with pinned in the end. Mostly on the basis that that's what the Android Contacts app does. It's also what we did before, and I'm going to assume that, despite the likelihood of it being a random choice, we did a detailed UX study and determined that there was no better default. ;-)

**Pull request merged**